src/blockchain.js

```diff
@@ -15,33 +17,75 @@ function getAccounts(cb) {
   web3.eth.getAccounts(cb);
 };
 
-let Blockchain = {};
-
 Blockchain.connect = function(connectionList, opts, doneCb) {
   const self = this;
-  this.web3 = null;
-  this.doFirst(function(cb) {
-    reduce(connectionList, '', function(prev, value, next) {
-      if (prev === false) {
-        return next(null, false);
+
+  const checkConnect = (next) => {
+    getAccounts(function(err, a) {
+      if (err) {
+        web3.setProvider(null);
       }
+      return next(null, !err)
+    });
+  }
 
-      if (value === '$WEB3' && (typeof web3 !== 'undefined' && typeof Web3 !== 'undefined')) {
-        web3.setProvider(web3.givenProvider);
-      } else if (value !== '$WEB3' && (typeof Web3 !== 'undefined' && ((typeof web3 === 'undefined') || (typeof web3 !== 'undefined' && (!web3.isConnected || (web3.isConnected && !web3.isConnected())))))) {
-        if (value.indexOf('ws://') >= 0) {
-          web3.setProvider(new Web3.providers.WebsocketProvider(value));
-        } else {
-          web3.setProvider(new Web3.providers.HttpProvider(value));
+  const connectWeb3 = (next) => {
+    let exited = false;
+    if (web3.givenProvider) {
+      web3.setProvider(web3.givenProvider);
+      return checkConnect(next);
+    }
+    window.addEventListener('message', ({ data } = msg) => {
+      if (data && data.type === 'ETHEREUM_PROVIDER_SUCCESS') {
+        if(exited) {
+          return console.warn("%cNote: The application might be in a corrupted state, please reload the page", "font-size: 2em");
         }
-      } else if (value === '$WEB3') {
-        return next(null, '');
+        exited = true;
+        web3.setProvider(ethereum);
+        checkConnect(next);
       }
+    });
+    window.postMessage({ type: 'ETHEREUM_PROVIDER_REQUEST' }, '*');
 
-      getAccounts(function(err, account) {
-        return next(null, !!err)
-      });
-    }, function(err, _result) {
+    const infiniteWaitBreaker = setInterval(() => {
+      clearInterval(infiniteWaitBreaker);
```

Comment:
This clear interval is useless, because you declare `infiniteWaitBreaker` in the same scope as the function that calls the interval. Either remove the clear or move `infiniteWaitBreaker` higher.

Reply:
The clearInterval was working fine, however you were right it was useless, using setTimeout is better